### PR TITLE
add plugins, set yosys install for plugin

### PR DESCRIPTION
**Pull request by coolbreeze413**
> ### Motivate of the pull request
> - [ ] To address an existing issue. If so, please provide a link to the issue: <issue id>
> - [ ] Breaking new feature. If so, please describe details in the description part.

Add support for yosys plugins in OpenFPGA (yosys-symbiflow-plugins)

> ### Describe the technical details
> #### What is currently done? (Provide issue link if applicable)
> 
> 
> 
> 
> 
Added yosys-symbiflow-plugins as a submodule
Set yosys paths according to expectation of yosys makefile ( using with plugins )

>
> #### What does this pull request change?
> 
> 
> 
> 
> 

> ### Which part of the code base require a change
> 
> - [ ] VPR
> - [ ] Tileable routing architecture generator
> - [ ] OpenFPGA libraries
> - [ ] FPGA-Verilog
> - [ ] FPGA-Bitstream
> - [ ] FPGA-SDC
> - [ ] FPGA-SPICE
> - [ ] Flow scripts
> - [ ] Architecture library
> - [ ] Cell library
> - [ ] Documentation
> - [ ] Regression tests
> - [ ] Continous Integration (CI) scripts

> ### Impact of the pull request

> - [ ] Require a change on Quality of Results (QoR)
> - [ ] Break back-compatibility. If so, please list who may be influenced.

### CODE REVIEW

File: CMakeLists.txt
Comment (arlpetergadfort) [inline CMakeLists.txt:234]: Why make clean? This will force every compile of OpenFPGA to fully recompile yosys instead of just the changes.  This adds significant time to the compile.

File: CMakeLists.txt
Comment (arlpetergadfort) [inline CMakeLists.txt:245]: Same as a above.

File: CMakeLists.txt
Comment (coolbreeze413) [reply CMakeLists.txt]: yes, this is correct, removed with https://github.com/lnis-uofu/OpenFPGA/pull/408/commits/a823c3e143ce6d67dd47d820b2eba29e498760b2

File: CMakeLists.txt
Comment (coolbreeze413) [reply CMakeLists.txt]: yes, this is correct, removed with https://github.com/lnis-uofu/OpenFPGA/pull/408/commits/a823c3e143ce6d67dd47d820b2eba29e498760b2

**Comment by coolbreeze413** [comment]
Removing the clean step in the CMakeLists.txt resulted in a strange error: https://github.com/lnis-uofu/OpenFPGA/runs/4101415984?check_suite_focus=true

I am not sure how, will debug further.

**Comment by coolbreeze413** [comment]
Looks like it might have been a temporary glitch, rerunning the jobs, the CI looks ok

**Comment by coolbreeze413** [comment]
As a reference for the modifications required for the yosys installation paths, this discussion in the older PR would be relevant: https://github.com/lnis-uofu/OpenFPGA/pull/371#issuecomment-941750875

**Comment by tangxifan** [comment]
@coolbreeze413 Code changes look clean, can you try to enable QL's tests at 

https://github.com/lnis-uofu/OpenFPGA/blob/a42342fa9e84ed7d644d2b0b39498aaaefc7a10e/openfpga_flow/regression_test_scripts/quicklogic_reg_test.sh#L11-L21


This is to validate that the plug-in is indeed working. Or if you have other simple test case, feel free to add it to the dedicated regression test.
For this PR, a simple test case is enough.

**Comment by tangxifan** [comment]
@ganeshgore Can you double check the docker-related code changes?

**Comment by ganeshgore** [comment]
Docker changes look good to me, thank you @coolbreeze413

**Comment by coolbreeze413** [comment]
@ganeshgore @tangxifan Thanks for reviewing the code changes.
I see issues in the quicklogic tests after enabling, I have added a single counter test for now in a new commit to this PR so that is easier for debugging.

**Comment by tangxifan** [comment]
@coolbreeze413 No worries. I see the counter test case failed. Do you want to debug it and finish it in this PR? Or you prefer to do it in a follow-up PR?

**Comment by coolbreeze413** [comment]
@tangxifan I would prefer to debug in this PR, at least one basic test case passing would be good. I will work with @tpagarani  and @lnsharma tomorrow to figure out the issue, and then take a call if we want to move the debug to a separate PR.
Does that work for you?

**Comment by tangxifan** [comment]
> @tangxifan I would prefer to debug in this PR, at least one basic test case passing would be good. I will work with @tpagarani and @lnsharma tomorrow to figure out the issue, and then take a call if we want to move the debug to a separate PR. Does that work for you?

Sure. No worries.

**Comment by coolbreeze413** [comment]
@tangxifan This error needs some changes in the yosys-symbiflow-plugins to prevent these kind of assertions. So, we would have to hold this PR until the plugins repo is updated, probably in a day or two.
(I believe that it has already been discussed that using a NDEBUG option to yosys compilation is not a preferred approach to fix this.)

**Comment by tangxifan** [comment]
@coolbreeze413 No worries. No rush here.

**Comment by lnsharma** [comment]
@tangxifan : I checked in the fix for yosys-plugin. However, now CI is failing during yosys checkout. If compile OpenFPGA using 'make all', the compilation and installation goes fine but in CI yosys checkout fails. Looks like CI is not compiling yosys but try to  do checkout on a particular tag. Can you please suggest what is the difference in CI Vs doing compilation via 'make all'?

**Comment by coolbreeze413** [comment]
looks like CI is ok now (previous failure in yosys compilation does not seem to be related to anything in the removed clean steps) - might be a temporary glitch last time.
The quicklogic tests now pass through the synth with the plugins, but throw up an error while running VPR, investigating this now.
Other tests continue to be ok.

**Comment by coolbreeze413** [comment]
@tangxifan The fixes to the plugins are in place, and the basic test enabled passes with the plugins enabled.
Please review the changes, and let me know if it looks ok to merge.

File: openfpga_flow/tasks/quicklogic_tests/flow_test/config/task.conf
Comment (tangxifan) [inline openfpga_flow/tasks/quicklogic_tests/flow_test/config/task.conf:61]: @coolbreeze413 O.k. For the basic test, you mean the counter is running fine, right?

**Review by tangxifan — APPROVED** [review]

**Comment by tangxifan** [comment]
@coolbreeze413 @tpagarani Waiting for your confirmation. I will click merge button then.

File: openfpga_flow/tasks/quicklogic_tests/flow_test/config/task.conf
Comment (coolbreeze413) [reply openfpga_flow/tasks/quicklogic_tests/flow_test/config/task.conf:61]: yes that is correct, the counter is running fine now

**Comment by tpagarani** [comment]
lgtm

**Pull request merged**
